test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -167,54 +207,314 @@
+          (str "did not get back the same instance from the third borrow"
+               "attempt as was returned from the first borrow attempt")))))
+
+(deftest pool-can-borrow-after-borrow-timed-out-during-lock
+  (testing (str "can do a timed borrow from pool after previous borrow "
+                "timed out while a write lock was held")
+    (let [pool (create-populated-pool 1)
+          borrow-1 (.borrowItem pool)
+          borrow-thread-started-2? (promise)
+          borrow-thread-started-3? (promise)
+          borrow-thread-2 (future
+                           (deliver borrow-thread-started-2? true)
+                           (.borrowItemWithTimeout
+                            pool
+                            50
+                            TimeUnit/MILLISECONDS))
+          borrow-thread-3 (future
+                           (deliver borrow-thread-started-3? true)
+                           (.borrowItemWithTimeout pool
+                                                   1
+                                                   TimeUnit/SECONDS))
+          _ @borrow-thread-started-2?
+          _ @borrow-thread-started-3?
+          unlock? (promise)
+          lock-thread-started? (promise)
+          lock-thread-locked? (promise)
+          lock-thread (future
+                       (deliver lock-thread-started? true)
+                       (.lock pool)
+                       (deliver lock-thread-locked? true)
+                       @unlock?
+                       (.unlock pool))]
+      @lock-thread-started?
+      (.releaseItem pool borrow-1)
+      @lock-thread-locked?
+      (is (true? true))
+      (is (nil? @borrow-thread-2)
+          "second borrow attempt did not time out")
+      (deliver unlock? true)
+      @lock-thread
+      (is (true? true))
+      (is (identical? borrow-1 @borrow-thread-3)
+          (str "did not get back the same instance from the third borrow"
+               "attempt as was returned from the first borrow attempt")))))
+
+(deftest pool-insert-pill-test
+  (testing "inserted pill is next item borrowed"
+    (let [pool (create-populated-pool 1)
+          pill (str "i'm a pill")]
+      (.insertPill pool pill)
+      (is (identical? (.borrowItem pool) pill)))))
+
+(deftest pool-clear-test
+  (testing (str "pool clear removes all elements from queue and only matching"
+                "registered elements")
+    (let [pool (create-populated-pool 3)
+          instance (.borrowItem pool)]
+      (is (= 2 (.size pool)))
+      (is (= 3 (-> pool
+                   (.getRegisteredElements)
+                   (.size))))
+      (.clear pool)
+      (is (= 0 (.size pool)))
+      (let [registered-elements (.getRegisteredElements pool)]
+        (is (= 1 (.size registered-elements)))
+        (is (identical? instance (-> registered-elements
+                                     (.iterator)
+                                     iterator-seq
+                                     first)))))))
```

Comment:
It makes sense.  It seems like the desired behavior would be that `clear` would have to block until it had freed all registered elements, but that is probably not worth dealing with right now.  I'm probably fine leaving this as is.  One other idea, though, might be to simply delete the `clear` method altogether for now?  I suspect that maybe the only reason it was there before was because we were directly inheriting from a queue class that already defined it, and we needed to override the behavior?